clusterloader/docs/vision.md

```diff
@@ -0,0 +1,344 @@
+# Cluster loader vision
+
+Author: wojtek-t
+
+Last update time: 9th May 2018
+
+## Background
+
+As of 31/03/2018, all our scalability tests, are regular e2e tests written in Go.
+It makes them really unfriendly for developers not working on scalability who just
+want to load test Kubernetes features they are working on. Doing so in many
+situations requires understanding how those tests really work, modify their code
+to test the new feature and only then run and debug tests. Alternatively, they
+may create a dedicated load test for their particular feature on their own, which
+may be easier, but on the other hand may not exercise important metrics that our
+performance tests check. This workflow is far from optimal.
+
+That said, long time ago we came up with the idea that users should be able to
+just bring their own objects definitions in JSON format, potentially annotate them
+with some metadata to describe how the load should be generated from these and
+testing infrastructure should do everything else automatically.
+
+In early 2017 the prototype of “Cluster Loader” has been created. It proved that
+configuring tests with json/yaml files is possible. But its functionality is very
+limited and it is very far from enabling migration any of existing scalability
+tests to that framework.
+
+We would like to get back to that and build fully functional Cluster Loader and
+use it as a framework to run all our scalability tests. This doc is describing
+high-level vision of how this will work.
+
+
+## Vision
+
+At the high level, a single test will consist of a number of steps. If we have
+really good usecase, we may consider creating a DAG of steps, but for now we
```

Comment:
Btw - I think this DAG argument can now be removed with the newer design. It's possible to achieve even a DAG of steps now with just a sequence of steps by merging nodes of the DAG by BFS level number. And by merging nodes, I mean joining their []Phases. Wdyt?

Reply:
Yeah - I think it no longer makes much sense. Removed.